src/accountsdb/buffer_pool.zig

```diff
@@ -0,0 +1,1074 @@
+
+        const freq = try allocator.alignedAlloc(u2, std.mem.page_size, num_frames);
+        errdefer allocator.free(freq);
+        @memset(freq, 0);
+
+        const in_queue = try allocator.alignedAlloc(InQueue, std.mem.page_size, num_frames);
+        errdefer allocator.free(in_queue);
+        @memset(in_queue, .none);
+
+        const size = try allocator.alignedAlloc(FrameOffset, std.mem.page_size, num_frames);
+        errdefer allocator.free(size);
+        @memset(size, 0);
+
+        return .{
+            .rc = rc,
+            .key = key,
+            .freq = freq,
+            .in_queue = in_queue,
+            .size = size,
+        };
+    }
+
+    fn deinit(self: *FramesMetadata, allocator: std.mem.Allocator) void {
+        // NOTE: this check itself is racy, but should never happen
+        for (self.rc) |*rc| {
+            if (rc.isAlive()) {
+                @panic("BufferPool deinitialised with alive handles");
+            }
+        }
+        allocator.free(self.rc);
+        allocator.free(self.key);
+        allocator.free(self.freq);
+        allocator.free(self.in_queue);
+        allocator.free(self.size);
+        self.* = undefined;
+    }
+
+    // to be called on the eviction of a frame
+    // should never be called on a frame with rc>0
+    // TODO: this should *all* be atomic (!)
+    fn resetFrame(self: FramesMetadata, index: FrameIndex) void {
+        if (self.rc[index].isAlive()) unreachable;
+        self.freq[index] = 0;
+        self.in_queue[index] = .none;
+        self.size[index] = 0;
+        self.key[index] = FileIdFileOffset.INVALID;
+    }
+};
+
+/// This cache is S3-FIFO inspired, with some important modifications to reads
+/// and eviction. This cache:
+/// 1) Does not store any values itself.
+/// 2) Never "forgets" any key that is inserted, meaning it stays inside the
+///    cache until eviction is called.
+/// 3) Maintains that main.cap and ghost.cap equal num_frames.
+/// 4) Asserts that its main and ghost queues never get full.
+/// 5) Does not allow duplicates, except when a ghost key is read again, leading
+///    to it pushed to main with a frequency of 1. The key remains in ghost,
+///    which is ignored when popped.
+/// 6) Checks frame refcounts upon eviction attempt, and will push alive frame
+///    indices back into the main queue (freq=1).
+/// 7) Asserts that it will always return a value upon eviction. Calling
+///    eviction when no free frames can be made available is illegal behaviour.
+pub const HierarchicalFIFO = struct {
+    pub const Key = FrameIndex;
+    pub const Metadata = FramesMetadata;
+    pub const Fifo = std.fifo.LinearFifo(FrameIndex, .Slice); // TODO: atomics
+
+    small: Fifo,
+    main: Fifo, // probably-alive items
+    ghost: Fifo, // probably-dead items
+
+    pub fn init(allocator: std.mem.Allocator, small_size: u32, num_frames: u32) !HierarchicalFIFO {
+        if (small_size > num_frames) unreachable;
```

Comment:
same

Reply:
downgraded to an error. https://github.com/Syndica/sig/pull/368/commits/1c8f4bdcfe4295dac4f14acabd0ae3af413c169e#diff-9334997573b7124ee92d8ed1efaa5a70b34f8531988349743900d1cf39014569L526-R558